### ISSUE TITLE
Arbitrary functions fail when the Gen instance on their return type sometimes fails

### ISSUE THREAD

**Issue by nrinaudo** [rickynils/scalacheck#300]
This means, among other things, that any `Arbitrary` instance that relies on `Gen.suchThat` cannot be used to generate arbitrary functions.

As an example:

```scala
import org.scalacheck._

implicit val nonZero: Arbitrary[Int] = Arbitrary(Arbitrary.arbitrary[Int].suchThat(_ != 0))
val prop = Prop.forAll { (f: String => Int) => f("foo"); true }
```

And then:

```
scala> prop.check
java.lang.StackOverflowError
  ...
```

See alexarchambault/scalacheck-shapeless#50 for a far more insidious example.

**Comment by non** [comment]
So, just to be clear, the generator has to fail many times (by default 100) before the function generator fails. See `pureApply` and `doPureApply` here: https://github.com/rickynils/scalacheck/blob/master/src/main/scala/org/scalacheck/Gen.scala#L73.

I think this is the only reasonable behavior for a generator that can be filtered (it is roughly the same behavior as the original QuickCheck's). By contrast, @xuwei-k's ScalaProps library does not even include `suchThat` or `filter`, which is a cleaner design (it removes the possibility of generator exhaustion during tests and also ensures function generators can't fail).

My sense is that ScalaCheck-Shapeless wasn't just failing some of the time, but a significant fraction of the time (e.g. 50% or more) which seems like it needed to be fixed anyway.

**Comment by alexarchambault** [comment]
@non Per [the source](https://github.com/rickynils/scalacheck/blob/b84b5a029f3d4f85f005f65c729877a989c0f48f/src/main/scala/org/scalacheck/Gen.scala#L55-L59), `doApply` is only called once.

**Comment by alexarchambault** [comment]
As @nrinaudo's example illustrate, it only takes one failure to make `Gen.function1` fail.

```scala
import org.scalacheck._

implicit val nonZero: Arbitrary[Int] = Arbitrary(Arbitrary.arbInt.arbitrary.suchThat(_ != 0))
val prop = Prop.forAll { (f: String => Int) => f("foo"); true }

prop.check
```
gives
```
! Exception raised on property evaluation.
> ARG_0: <function1>
> Exception: org.scalacheck.Gen$RetrievalError: couldn't generate value
org.scalacheck.Gen.loop$1(Gen.scala:57)
org.scalacheck.Gen.doPureApply(Gen.scala:58)
org.scalacheck.Gen.pureApply(Gen.scala:72)
org.scalacheck.GenArities$$anonfun$function1$1$$anonfun$1.apply(GenArities.
  scala:15)
```

**Comment by non** [comment]
OK! Yes, there is a bug/typo, the method should read:

```scala
  def doPureApply(p: Gen.Parameters, seed: Seed, retries: Int = 100): Gen.R[T] = {
    @tailrec def loop(r: Gen.R[T], i: Int): Gen.R[T] = {
      if (r.retrieve.isDefined) r
      else if (i > 0) loop(doApply(p, r.seed), i - 1)
      else throw new Gen.RetrievalError()
    }
    loop(doApply(p, seed), retries)
  }
```

I've confirmed that with this change the property passes.

**Comment by peter-empen** [comment]
I'd appreciate your publishing this fix for Scala 2.11. Thanks a lot.
